cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +119,79 @@ public Void call() throws Exception {
     }
   }
 
+  private void cleanFilesWithoutMeta(final long tillTime, final Map<String, NamespaceId> namespacedLogBaseDirMap,
+                                     final SetMultimap<String, Location> parentDirs) throws Exception {
+    List<NamespaceMeta> namespaceMetaList = namespaceQueryAdmin.list();
+    for (NamespaceMeta namespaceMeta : namespaceMetaList) {
+      final NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      final String namespacedLogBaseDir = LoggingContextHelper.getNamespacedBaseDir(namespacedLocationFactory,
+                                                                                    logBaseDir, namespaceId);
+      Location startLocation = rootLocationFactory.create(namespacedLogBaseDir);
+      if (startLocation.exists()) {
+        Processor<LocationStatus, Set<Location>> processor = getLocationProcessor(tillTime);
+        Locations.processLocations(startLocation, true, processor);
+
+        Set<Location> result = processor.getResult();
+        Set<Location> scanLocations = fileMetaDataManager.scanFiles(namespaceId.getNamespace(), tillTime);
+
+        for (final Location location : Sets.difference(result, scanLocations)) {
+          LOG.debug("Log file {} does not have corresponding meta data.", location.toString());
```

Comment:
Or, since we are already logging every file delete at info level, I think it is okay if you add one log line like "Starting deletion of log files older than $time without metadata" at the beginning of `cleanFilesWithoutMeta`.